addons/knowledge/static/src/js/knowledge_behavior_table_of_content.js

```diff
@@ -0,0 +1,214 @@
+        this._super.apply(this, arguments);
+        if (this.mode === 'edit') {
+            this.anchor.querySelectorAll('a').forEach(element => {
+                element.setAttribute('contenteditable', 'false');
+            });
+        }
+    },
+
+    /**
+     * Adds the ToC click listener to scroll towards to associated heading tag.
+     * Also adds the listener that will update the ToC as the user is typing.
+     *
+     * @override
+     */
+    applyListeners: function () {
+        this._super.apply(this, arguments);
+        $(this.anchor).on('click', '.o_knowledge_toc_link', this._onTocLinkClick.bind(this));
+        if (this.mode === 'edit') {
+            this.observer.observe(this.handler.field, {
+                childList: true,
+                attributes: false,
+                subtree: true,
+                characterData: true,
+            });
+
+            this._updateTableOfContents();
+        }
+    },
+
+    /**
+     * @override
+     */
+    disableListeners: function () {
+        $(this.anchor).off('click', '.o_knowledge_toc_link');
+        if (this.mode === 'edit') {
+            this.observer.disconnect();
+        }
+    },
+
+    //--------------------------------------------------------------------------
+    // Table of content - BUSINESS LOGIC
+    //--------------------------------------------------------------------------
+
+    /**
+     * Allows to debounce the update of the Table of Content to avoid updating whenever every single
+     * character is typed. The debounce is set to 500ms.
+     */
+    delayedUpdateTableOfContents() {
+        if (this.updateTimeout) {
+            clearTimeout(this.updateTimeout);
+        }
+        this.updateTimeout = setTimeout(this._updateTableOfContents.bind(this), 500);
+    },
+
+    /**
+     * Updates the Table of Content to match the document headings.
+     * We pause our observer during the process.
+     */
+    _updateTableOfContents: function () {
+        this.handler.editor.observerUnactive('knowledge_toc_update');
+
+        const allHeadings = fetchRelevantHeadings(this.handler.field);
+
+        let currentDepthByTag = {};
+        let previousTag = undefined;
+        let previousDepth = -1;
+        let index = 0;
+        const headingStructure = allHeadings.map((heading) => {
+            let depth = HEADINGS.indexOf(heading.tagName)
+            // Compute the 'depth' we want to display when increasing
+            if (depth !== previousDepth && heading.tagName === previousTag) {
+                // Same tag name as previous one -> use same depth
+                depth = previousDepth;
+            } else if (depth > previousDepth) {
+                // This is done only when changing tags, as 2 h4 in our previous example
+                // need to be at the same depth
+                if (heading.tagName !== previousTag) {
+                    // We only go max +1 depth at the time, meaning if a h4 follows a h1,
```

Comment:
From:
```html
<h1>
    <h3>
```
Are we sure we want to display:
```html
<h1>
    <h3>
    <h2>
```
and not
```html
<h1>
        <h3>
    <h2>
```
?

Reply:
I think it is what we want yes, from what I understand from the specifications.

(And otherwise it would make the logic a lot more complicated).